### DISCUSSION TITLE
Compression avif bad on newer server

### DISCUSSION THREAD

**Discussion by tag6** [ImageMagick/ImageMagick#4411] [Help]
Hello,

currently I'm in the process of switching from an older server currently still running Ubuntu 14.04 to a newer one which runs Ubuntu 20.04. I'm using ImageMagick  which I compiled myself for better support of some formats, especially AVIF. But when I run the compression, on the new server the avif result does by far not get as small as on the old one.

Example:

`wget https://peach.blender.org/wp-content/uploads/poster_bunny_small.jpg' --> size 64432

'convert poster_bunny_small.jpg -auto-orient -strip -gaussian-blur 0.01 -quality 37% -resize 640x640 poster_bunny_small.jpg.mini.avif`

Result on the old server is 2407 bytes. But on the new server it takes 8049 bytes. On both servers libde265 and libheif are compiled. identify -list format gives AVIF* rw+ on both. Attached are the results as ZIP (this forum won't let my upload avif directly!).

If I change the file extension to generate from avif to jpg, the picture gets 8537 bytes. So it seems not the case that there is the jpg generated and stored as avif, what I understood would happen if the encoder would not work. But it is not much less in size, and it seems to always be in that range.

Has someone an idea what could be wrong? What am I missing? libheif has aom encoder yes in several places in config.log, and ImageMagick is configured --with-libheif=yes before compile.

Thank you for a helpful hint.

[poster_bunny_small.jpg.mini.u.zip](https://github.com/ImageMagick/ImageMagick/files/7413568/poster_bunny_small.jpg.mini.u.zip)

**Comment by fmw42** [comment]
Did the versions of libheif and libde265 change?

**Comment by tag6** [comment]
Yes, libde265 is 1.0.0 on the older server, on the newer one it is 1.0.4. libheif is same version. But on the newer server it seems there is a second libheif, an older one, not only the one that I compiled and make install. Remove the apt-installed version it does not find libheif anymore, make install and ldconfig still does not make it visible for dpkg -S liebheif ... hm but in any case the right question to find the cause, thank you!

**Comment by tag6** [comment]
Trying to find some solution to my problem, how to make sure the right libheif gets used, the hint is always "check (magick_component) -version and if you see delegate heic, everything is fine". That is a bad thing. Why does Imagemagick report it is available, but then it is not working right? There is something broken. It should fail if it does not have the right algorithm available. Why is avif not listed in the delegates? Why does Imagemagick report full read/write support rw+ for AVIF in "identify -list format", and then produces some nonsense as file? What am I missing? How can I get the information, which algorithm is being used by ImageMagick when encoding AVIF?

**Comment by tag6** [comment]
The problem for sure is in heiflib. heif-enc -A already gives files that are way too big.

**Comment by tag6** [comment]
On both systems: convert from avif to heif fails and produces a zero-byte-file.

convert poster_bunny_small.avif poster_bunny_small.heif
convert: Unsupported file-type: Unspecified `poster_bunny_small.heif' @ error/heic.c/IsHeifSuccess/135.

**Comment by fmw42** [comment]
What version of Imagemagick? Do you have the latest libheif and installed it without errors with all its needed delegates?

**Comment by tag6** [comment]
ImageMagick is 7.1.0.9 on the Ubuntu 14 where it works fine and 7.1.0.10 on Ubuntu 20.04 where it produces big files. Libheif is current release 1.12.0 on both servers, I tried trunk master on the newer server but reverted as it did not change anything. libde265 is 1.0.0 on the older server, on the newer one it is 1.0.4.

The problem that I see is that you see in ImageMagick delegates which make really no statement on installed codecs. Trying to convert an avif to heif using heif-enc gives on both server the error 'No HEVC encoder available.' - how can I ask ImageMagick if the image codec HEVC is available? It produces heif while heif-enc refuses to do so. Probably ImageMagick is using a worse algorithm than HEVC and still generating a heif-compliant file. For some image format there could be 10 codecs available. They are installed and working if they are not providing the best algorithm (which is what I am searching for). Why is HEVC not available? On the newer server libx265-dev is installed and should be working. There is no information, no warning, no error about this. This is wrong! It should state "if you want this working, make lib... work and re-compile and install". And this should happen when compiling AND when producing sup-optimal output (at least when setting some verbosity flag).

You can ask using identify -list format which formats are supported, but not single algorithms. For instance I had long time receiving my fax as tif - but a seldom used encoding (Fax-Group-4) was being used for this. Often the decoding was not supported. What is the command for ImageMagick to not only show the delegates and supported file extensions, but the available codecs for encoding and decoding? Can it read tiff? If it can, does it support the fax codec? Some other seldom implemented codec? Encoding also? How do I make ImageMagick tell that the best algorithm is not available, and the information about why and how to change this? How can I request ImageMagick to encode a file for some format using a special codec eg fax for tiff?

It is fine the ImageMagick relies on delegates. But it should manage a list of available variant for every delegate  - minimal working, maximal working, best compression, delegate-special strings (like FAX) - and it should provide a list of codecs. An interface where it can call a delegate "what codecs do you support?" and report this to the user. It is fine that then it would report "dunno" for every delegate in first place until this information is filled, but something like this will be needed.

Delegates ImageMagick on 14.04:
Delegates (built-in): freetype heic jng jpeg png webp x xml zlib

Delegates ImageMagick on 20.04:
Delegates (built-in): bzlib djvu fontconfig freetype heic jbig jng jp2 jpeg lcms lqr lzma openexr pangocairo png tiff webp x xml zlib

On both servers:
$ identify -list format | grep -i avif
     AVIF* rw+   AV1 Image File Format (1.12.0)

But it must be something inside libheif, see:

heif-enc -A poster_bunny_small.jpg -A -o poster_bunny_small.avif

On server Ubuntu 14.04 result:
7754 Okt 26 14:52 poster_bunny_small.avif

On Server Ubuntu 20.04 result:
45313 Oct 26 15:26 poster_bunny_small.avif

**Comment by tag6** [comment]
It is not a problem of libheif. Compiled [go-avif](https://github.com/Kagami/go-avif), used cmd from there and also got output that is way too big compared with the static build on release page of go-avif. So the cause of the big files clearly must be in libaom on the new server.

**Comment by tag6** [comment]
Did apt remove libaom. Compiled it myself. Re-built and make install heif-enc. Result convert for poster_bunny_small.jpg.mini.avif
 is now 2362 Bytes, wonderful. It is not good that Ubuntu 20.10 defaults to a really bad old libaom. My opinion: It should exist a minimum required version and libheif should refuse to compile if it is so old, that the resulting images are more than 10% bigger than those from current version. Maybe the compile of ImageMagick with libheif should in the future also restrict supporting delegate heif/avif to some minimum version.
